File: Data/Examples/Android/McAndroidTester/app/src/main/java/com/elbit/mapcore/mcandroidtester/managers/MapWorld/Manager_MCTerrain.java
package com.elbit.mapcore.mcandroidtester.managers.MapWorld;

import android.content.Context;

import com.elbit.mapcore.mcandroidtester.utils.AlertMessages;
import com.elbit.mapcore.mcandroidtester.utils.BaseApplication;

import java.util.HashMap;

import com.elbit.mapcore.General.MapCoreException;
import com.elbit.mapcore.Interfaces.Calculations.GridCoordSystem.IMcGridCoordinateSystem;
import com.elbit.mapcore.Interfaces.Map.IMcMapLayer;
import com.elbit.mapcore.Interfaces.Map.IMcMapTerrain;
import com.elbit.mapcore.mcandroidtester.utils.Funcs;

/**
 * Created by tc99382 on 02/08/2016.
 */
public class Manager_MCTerrain {

    static int currentTerrainID = 0;
    private static Manager_MCTerrain instance;

    private HashMap<Object, Integer> mdTerrain;

    public HashMap<Object, Integer> getTerrains() {
        return mdTerrain;
    }

    private Manager_MCTerrain()
    {
        mdTerrain = new HashMap<>();
    }

    public IMcMapTerrain CreateTerrain(IMcGridCoordinateSystem gridCoordSys, IMcMapLayer[] layerList, Context context)
    {
        IMcMapTerrain ret = null;
        try
        {
            ret = IMcMapTerrain.Static.Create(gridCoordSys, layerList);
            for (IMcMapLayer layer: layerList)
                Manager_MCLayers.getInstance().removeStandaloneLayer(layer);
            if (ret != null) {
                AddTerrain(ret);
            }
            else {
                AlertMessages.ShowErrorMessage(context, "Terrain creation", "Terrain creation failed!!!");
            }
        }
        catch (MapCoreException McEx)
        {
            AlertMessages.ShowMapCoreErrorMessage(context, McEx, "McMapTerrain.Create");
            McEx.printStackTrace();
        }
       catch (Exception e) {
            e.printStackTrace();
        }
        return ret;
    }

    public void AddTerrain(IMcMapTerrain terrain)
    {
        mdTerrain.put(terrain,currentTerrainID++);
    }

    public void RemoveTerrain(final IMcMapTerrain terrain) {
        mdTerrain.remove(terrain);
        Funcs.runMapCoreFunc(new Runnable() {
            @Override
            public void run() {
                terrain.Release();
            }
        });
    }

    public HashMap<Object, Integer> getAllParams() {
       return mdTerrain;
    }

    public HashMap<Object, Integer> GetChildren(Object Parent)
    {
        HashMap<Object, Integer> Ret = new HashMap<>();
        if (Parent instanceof IMcMapTerrain)
        {
            IMcMapTerrain Terrain = (IMcMapTerrain)Parent;
            IMcMapLayer[] LayersInTerrain;
            try {
                LayersInTerrain = Terrain.GetLayers();
                for (int i=0;i<=LayersInTerrain.length;i++)
                {
                    Ret.put(LayersInTerrain[i], i++);
                }
            } catch (MapCoreException McEx) {
                McEx.printStackTrace();
                AlertMessages.ShowMapCoreErrorMessage(BaseApplication.getCurrActivityContext(),McEx,"IMcMapLayer[] GetLayers()");
            }
            catch (Exception e) {
                e.printStackTrace();
            }
        }
        return Ret;
    }

    public static Manager_MCTerrain getInstance() {
        if (instance == null) {
            instance = new Manager_MCTerrain();
        }
        return instance;
    }
}

File: Data/Examples/CSharpTester/MCTester/GUI/Managers/MapWorld/Manager_MCTerrain.cs
using System;
using System.Collections.Generic;
using System.Text;
using MapCore;
using UnmanagedWrapper;
using MCTester.MapWorld;
using System.Windows.Forms;

namespace MCTester.Managers.MapWorld
{
    public static class Manager_MCTerrain
    {
        static uint currentTerrainID = 0;
        static Dictionary<IDNMcMapTerrain, uint> m_dTerrain;

        static Manager_MCTerrain()
        {
            m_dTerrain = new Dictionary<IDNMcMapTerrain, uint>();

        }

        public static IDNMcMapTerrain CreateTerrain(IDNMcGridCoordinateSystem gridCoordSys, IDNMcMapLayer[] layerList, bool bDisplayItemsAttachedTo3DModelWithoutDtm = false)
        {
            IDNMcMapTerrain ret = null;

            try
            {
                ret = DNMcMapTerrain.Create(gridCoordSys, layerList,null, null, bDisplayItemsAttachedTo3DModelWithoutDtm);

                if (ret != null)
                    dTerrain.Add(ret, currentTerrainID++);
                else
                    MessageBox.Show("Terrain creation failed!!!","Terrain creation",MessageBoxButtons.OK,MessageBoxIcon.Information);
            }
            catch (MapCoreException McEx)
            {
                MapCore.Common.Utilities.ShowErrorMessage("DNMcMapTerrain.Create", McEx);
            }
            
            return ret;
        }

        public static void AddTerrain(IDNMcMapTerrain terrain)
        {
            dTerrain.Add(terrain,currentTerrainID++);
        }

        public static void RemoveTerrainFromDic(IDNMcMapTerrain terrain)
        {
            if(dTerrain.ContainsKey(terrain))
                dTerrain.Remove(terrain);
            terrain = null;
        }

        public static void RemoveTerrain(IDNMcMapTerrain terrain)
        {
            if (dTerrain.ContainsKey(terrain))
                dTerrain.Remove(terrain);

            try
            {
                terrain.Dispose();
            }
            catch (MapCoreException McEx)
            {
                MapCore.Common.Utilities.ShowErrorMessage("Dispose terrain", McEx);
            }
            terrain = null;
        }
        

        public static  Dictionary<object, uint > AllParams
        {
            get
            {
                Dictionary<object, uint> Ret = new Dictionary<object, uint>();

                foreach (IDNMcMapTerrain keyTerr in dTerrain.Keys)
                {
                    Ret.Add(keyTerr, dTerrain[keyTerr]);
                }

                return Ret;
            }            
        }

        public static List<IDNMcMapTerrain> AllTerrains()
        {
            List<IDNMcMapTerrain> lst = new List<IDNMcMapTerrain>();
            foreach (IDNMcMapViewport vp in Manager_MCViewports.AllParams.Keys)
            {
                IDNMcMapTerrain[] Terrs = vp.GetTerrains();
                foreach (IDNMcMapTerrain terrain in Terrs)
                {
                    if (!lst.Contains(terrain))
                        lst.Add(terrain);
                }
            }

            foreach (IDNMcMapTerrain terr in AllParams.Keys)
            {
                if (!lst.Contains(terr))
                    lst.Add(terr);
            }
            return lst;
        }

        public static  Dictionary<object, uint > GetChildren(object Parent)
        {
            Dictionary<object, uint > Ret = new Dictionary<object, uint >();
            if (Parent is IDNMcMapTerrain)
            {
                IDNMcMapTerrain Terrain = (IDNMcMapTerrain)Parent;
                IDNMcMapLayer[] LayersInTerrain = Terrain.GetLayers();

                uint i = 0;
                foreach (IDNMcMapLayer currLayer in LayersInTerrain)
                {
                    if(!Ret.ContainsKey(currLayer))
                        Ret.Add(currLayer, i++);
                }
            }
            return Ret;
        }

        public static Dictionary<IDNMcMapTerrain, uint> dTerrain
        {
            get { return m_dTerrain; }
            set { m_dTerrain = value; }
        }
    }
}
